src/main/java/InfrastructureManager/EdgeClientHistory.java

```diff
@@ -0,0 +1,163 @@
+package InfrastructureManager;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+/**
+ * This class is dedicated to fetch history related info.
+ * From get history score, to history time to calculate the score.
+ * Implemented function : get historyScore - get ConnectedTime
+ * Also set historyMap and set ConnectedMap
+ */
+public class EdgeClientHistory {
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    // - -------------------------------- ---- THIS IS THE CONSTRUCTOR -----------------------------
+    public EdgeClientHistory(String id) {
+        this.id = id;
+    }
+
+    //put client ID in here, put client's connected node and node's score.
+    //need to get time where client's connected to node
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    public Multimap<String, HashMap<String, Long>> getHistoryMap() {
+        return historyMap;
+    }
+
+    /**
+     * Assign client, node and history score into class's multimap.
+     * This function create a new HashMap, use mainly for initialize out collection.
+     * Don't use to replace value, because we are using multimap, 1 key can have several similar value
+     * @param clientID
+     * @param nodeID
+     * @param nodeHistory
+     */
+    public void setHistoryMap(String clientID, String nodeID, Long nodeHistory) {
+        HashMap<String, Long> nodeAndScore = new HashMap<>();
+        nodeAndScore.put(nodeID, nodeHistory);
+        this.historyMap.put(clientID, nodeAndScore);
+    }
+
+    public void setHistoryScoreForClient(String clientID, String nodeID, Long nodeHistoryScore){
+        Collection<HashMap<String, Long>> clientHistoryMapList = this.historyMap.get(clientID);
+        for (HashMap<String, Long> theList : clientHistoryMapList){
+            if (theList.containsKey(nodeID)){
+                //replace with new score
+                theList.put(nodeID,nodeHistoryScore);
+                break;
+            }
+        }
+
+    }
+
+    /**
+     * Assign client, node and connected time into class's multimap.
+     * This function create a new HashMap, use mainly for initialize out collection.
+     * Don't use to replace value, because we are using multimap, 1 key can have several similar value
+     *
+     * @param clientID
+     * @param nodeID
+     * @param connectedTime
+     */
+    public void setConnectedMap(String clientID, String nodeID, Long connectedTime) {
```

Comment:
I use this to populate the map with all nodes in nodeList and connectedTime =0 when client is registered. 
set connectedTime only set the time, not the node. 
Sure there is better way (or more compact way) to do it but seperate into different function help me reduce the complexity of the class and the relevant

Reply:
You can keep them separate but only expose the functions that make sense from a black box point of view. I would say it is more complicated that I need to call this function before the other one otherwise it doesn't work. The user of a class should use the class in a logical way without knowing the implementation of the class. And here as a user of this class I just need to set a connected time for a pair of client and node, I don't care if it needs initialization or not, and I certainly don't need to call two different functions to do so.

Reply:
done in 4364838